app/src/main/java/io/apicurio/registry/ibmcompat/api/impl/ApiServiceImpl.java

```diff
@@ -52,12 +77,147 @@
+            schemaVersion.setName(artifactMetaData.getName());
+            schemaVersion.setEnabled(!ArtifactState.DISABLED.equals(artifactMetaData.getState()));
+            SchemaState versionState = new SchemaState();
+            if(ArtifactState.DEPRECATED.equals(artifactMetaData.getState())) {
+                versionState.setState(SchemaState.StateEnum.DEPRECATED);
+            } else {
+                versionState.setState(SchemaState.StateEnum.ACTIVE);
+            }
+            schemaVersion.setState(versionState);
+        } catch (ArtifactNotFoundException e) {
+            return null;
+        }
+        return schemaVersion;
+    }
+
+    private void populateSchemaSummary(String schemaid, SchemaSummary schemaSummary) {
+        List<ArtifactState> versionStates = storage.getArtifactVersions(schemaid).stream()
+            .map(version -> storage.getArtifactVersionMetaData(schemaid, version).getState())
+            .collect(Collectors.toList());
+
+        schemaSummary.setId(schemaid);
+        schemaSummary.setName(schemaid); // TODO - add mechanism to store an artifact-level name
+
+        // The schema is disabled if all versions are disabled
+        boolean isSchemaDisabled = versionStates.stream().allMatch(versionState -> ArtifactState.DISABLED.equals(versionState));
+        schemaSummary.setEnabled(!isSchemaDisabled);
+
+        // The schema is deprecated if all versions are deprecated
+        boolean isSchemaDeprecated = versionStates.stream().allMatch(versionState -> ArtifactState.DEPRECATED.equals(versionState));
+        SchemaState schemaState = new SchemaState();
+        if(isSchemaDeprecated) {
+            schemaState.setState(SchemaState.StateEnum.DEPRECATED);
+        } else {
+            schemaState.setState(SchemaState.StateEnum.ACTIVE);
+        }
+        schemaSummary.setState(schemaState);
+    }
+
+    private void handleArtifactCreation(AsyncResponse response, String artifactId, String versionName, Throwable t) {
+        if (t != null) {
+            response.resume(t);
+            return;
+        }
+
+        if (!doesArtifactExist(artifactId)) {
+            // Some storage (such as asyncmem) needs a wait before the artifact is available for updating
+            try {
+                waitForArtifactCreation(artifactId);
+            } catch (ArtifactNotFoundException e) {
+                response.resume(e);
+                return;
+            }
+        }
+
+        try {
+            // Set the artifact name from the version name
+            EditableArtifactMetaDataDto dto = new EditableArtifactMetaDataDto();
+            dto.setName(versionName);
+            //
+            // TODO: On Kafka topic storage, this call to updateArtifactMetaData never returns.
+            // The consumer thread appears to be blocked. Help needed here.
+            //
+            // storage.updateArtifactMetaData(artifactId, dto);
```

Comment:
Right - this will not work. @alesj can provide a good explanation.  This has come up before and I was looking for a link to that explanation but I couldn't dig it up quickly.

Reply:
A single (separate) call should not be a problem.
We had problems were we nested ComplationStage calls -- and the thread of the first one was used for others as well. Of course then the first one never finished.

Reply:
Is that not what's happening here?  I assumed it was the same problem.

Reply:
I guess it could be ... if the method that calls this is called in some nested CS -- didn't track that much :-)